csrc/device_lower/pass/allocation.cpp

```diff
@@ -163,11 +163,12 @@ class AllocationInserter : public kir::ExprMutator {
     std::vector<Val*> alloc_dims;
 
     for (const auto id : maybe_rfactor_domain) {
-      if (id->isReduction() || id->isStride() || id->isDeviceDim()) {
+      if (id->isReduction() || id->isStride()) {
         continue;
-      } else if (id->isBroadcast()) {
+      } else if (id->isBroadcast() || id->isDeviceDim()) {
```

Comment:
The change makes sense to me, but I also wonder if this should be also done with the other IterTypes, i.e., `isReduction()` and `isStride()`. For example, if a tensor only has a reduction domain, it seems that means it won't have any allocation dimension, which seems wrong. The existing schedulers should never put reduction outputs on global memory, so it may be that we haven't seen any issue with this. 

Unless I missed anything, could you try creating a simple repro and see if this is a real issue?

Reply:
Are there coverage metrics that can check if line 166 gets executed?

Reply:
I ran the nvfuser_tests locally and the if branch on 166 is not triggered in any of the single device tests, which aligns with your assumption that the existing schedulers never put reductions on global memory. 

It does get executed in the multidevice_tests. Changing  `isReduction()` and `isStride()` axes to have an allocation size of 1 causes the tests to fail. 

We need to look into whether it's necessary for sharded tensors or if it will be fixed with proper memory allocation since the multidevice executor puts all TVs in global memory. (https://github.com/NVIDIA/Fuser/blob/main/csrc/multidevice/executor.cpp#L46)

Reply:
C++ exception with description "Expected T1_g[ bS4{1}, iS5{2}, iS6{3}, rS7{5} ] (DeviceMesh{0, 2, 3, }) to be bound to a tensor of rank 3, but got a tensor of rank 4
Exception raised from validateValWithConcreteValue at Fuser/csrc/expr_evaluator.cpp:38

Reply:
Ah, I see. That makes sense. I thought here it's just calculating the allocation size, but it's actually the shape of the allocation, so for reductions there should be no domain at all. Sorry for the noise.